tools/k8s_versions/update_k8s_versions.go

```diff
@@ -0,0 +1,222 @@
+			return nil, fmt.Errorf("error parsing date: %w", err)
+		}
+
+		for i, line := range versionBlock {
+			trimmed := strings.TrimSpace(line)
+			tag := ""
+			if strings.HasPrefix(trimmed, fmt.Sprintf("- v%s", version.Cycle)) {
+				if eolDate.After(now) {
+					tag, err = getAvailableKindImage(version.Latest)
+					if err != nil {
+						return nil, fmt.Errorf("failed to get available kind image: %w", err)
+					}
+					line = versionRE.ReplaceAllString(line, tag)
+					output = append(output, line)
+				}
+				break
+			} else if i == len(versionBlock)-1 && eolDate.After(now) {
+				// Add the new version at the end of the list
+				tag, err = getAvailableKindImage(version.Latest)
+				if err != nil {
+					return nil, fmt.Errorf("failed to get available kind image: %w", err)
+				}
+				output = append(output, fmt.Sprintf("%s- v%s # EOL %s", strings.Repeat(" ", len(line)-len(trimmed)), tag, version.EOLDate))
+			}
+		}
+	}
+	return output, nil
+}
+
+func getAvailableKindImage(tag string) (string, error) {
+	for {
+		exists, err := imageTagExists(tag)
+		if err != nil {
+			return "", fmt.Errorf("failed to check image tag existence: %w", err)
+		}
+		if exists {
+			return tag, nil
+		}
+
+		tag, err = decrementMinorMinorVersion(tag)
+		if err != nil {
+			return "", fmt.Errorf("failed to decrement version: %w", err)
+		}
+	}
+}
+
+func imageTagExists(tag string) (bool, error) {
+	url := "https://hub.docker.com/v2/repositories/kindest/node/tags?page_size=1&page=1&ordering=last_updated&name="
+	resp, err := http.Get(url + tag)
+	if err != nil {
+		return false, fmt.Errorf("failed getting k8s versions: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return false, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return false, fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	var kindImage DockerImage
+	if err := json.Unmarshal(body, &kindImage); err != nil {
+		return false, fmt.Errorf("failed to unmarshal JSON: %w", err)
+	}
+
+	if kindImage.Count > 0 {
+		return true, nil
+	}
+	return false, nil
+}
+
+func decrementMinorMinorVersion(version string) (string, error) {
+	parts := strings.Split(version, ".")
+	if len(parts) < 3 {
+		return "", fmt.Errorf("version does not have a minor version: %s", version)
+	}
+
+	minor, err := strconv.Atoi(parts[2])
+	if err != nil {
+		return "", fmt.Errorf("invalid minor version: %s", parts[1])
+	}
+
+	if minor == 0 {
+		return "", fmt.Errorf("minor version cannot be decremented below 0")
+	}
+
+	parts[2] = strconv.Itoa(minor - 1)
+	return strings.Join(parts, "."), nil
+}
+
+func main() {
+	url := "https://endoflife.date/api/kubernetes.json"
+	k8sVersions, err := fetchKubernetesVersions(url)
+	if err != nil {
+		fmt.Println("Failed to get k8s versions:", err)
+		return
+	}
+
+	filesToUpdate := []string{".github/workflows/functional_test.yaml", ".github/workflows/functional_test_v2.yaml"}
```

Comment:
nit: nice to have these as parameters, but, it can be as it is too

Reply:
I am planning on expanding this to include EKS, AKS and GKE with an option to choose what cloud provider list to update and I will also add the option to pass the list of yaml file to update and the name of the array element where the versions exist. 
But this will be in the next version